oadp-operator/controllers/restic.go

```diff
@@ -1,14 +1,263 @@
+		// Deployment selector is immutable so we set this value only if
+		// a new object is going to be created
+		if ds.ObjectMeta.CreationTimestamp.IsZero() {
+			ds.Spec.Selector = &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"component": veleroSAName,
+				},
+			}
+		}
+
+		if err := controllerutil.SetControllerReference(&velero, ds, r.Scheme); err != nil {
+			return err
+		}
+
+		if _, err := r.buildResticDaemonset(&velero, ds); err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err != nil {
+		return false, err
+	}
+
+	if op == controllerutil.OperationResultCreated || op == controllerutil.OperationResultUpdated {
+		// Trigger event to indicate restic was created or updated
+		r.EventRecorder.Event(ds,
+			v1.EventTypeNormal,
+			"ResticDaemonsetReconciled",
+			fmt.Sprintf("performed %s on restic deployment %s/%s", op, ds.Namespace, ds.Name),
+		)
+	}
+
 	return true, nil
 }
+
+/**
+ * This function builds restic Daemonset. It calls /pkg/credentials function AppendCloudProviderVolumes
+ * args: velero - the velero object pointer
+ * 		 ds		- pointer to daemonset with objectMeta defined
+ * returns: (pointer to daemonset, nil) if successful
+ */
+func (r *VeleroReconciler) buildResticDaemonset(velero *oadpv1alpha1.Velero, ds *appsv1.DaemonSet) (*appsv1.DaemonSet, error) {
+	if velero == nil {
+		return nil, fmt.Errorf("velero cannot be nil")
+	}
+	if ds == nil {
+		return nil, fmt.Errorf("ds cannot be nil")
+	}
+	ds.Spec = appsv1.DaemonSetSpec{
+		Selector: &metav1.LabelSelector{
+			MatchLabels: resticLabelMap,
+		},
+		UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+			Type: appsv1.RollingUpdateDaemonSetStrategyType,
+		},
+		Template: v1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: resticLabelMap,
+			},
+			Spec: v1.PodSpec{
+				NodeSelector:       velero.Spec.ResticNodeSelector,
+				ServiceAccountName: veleroSAName,
+				SecurityContext: &v1.PodSecurityContext{
+					RunAsUser:          pointer.Int64(0),
+					SupplementalGroups: []int64{},
+				},
+				Volumes: []v1.Volume{
+					// Cloud Provider volumes are dynamically added in the for loop below
+					{
+						Name: "host-pods",
+						VolumeSource: v1.VolumeSource{
+							HostPath: &v1.HostPathVolumeSource{
+								Path: resticPvHostPath,
+							},
+						},
+					},
+					{
+						Name: "scratch",
+						VolumeSource: v1.VolumeSource{
+							EmptyDir: &v1.EmptyDirVolumeSource{},
+						},
+					},
+					{
+						Name: "certs",
+						VolumeSource: v1.VolumeSource{
+							EmptyDir: &v1.EmptyDirVolumeSource{},
+						},
+					},
+				},
+				Tolerations: velero.Spec.ResticTolerations,
+				Containers: []v1.Container{
+					{
+						Name: "velero",
+						SecurityContext: &v1.SecurityContext{
+							Privileged: pointer.Bool(true),
+						},
+						Image: fmt.Sprintf("%v/%v/%v:%v", os.Getenv("REGISTRY"), os.Getenv("PROJECT"), os.Getenv("VELERO_REPO"), os.Getenv("VELERO_TAG")),
```

Comment:
Let's create a helper image `getResticImage()` that builds this out of the env vars rather than putting it all on one line